### DISCUSSION TITLE
Need help with connecting my Go script with PHP side through RPC

### DISCUSSION THREAD

**Discussion by PakToPack** [roadrunner-server/roadrunner#1612] [Q&A]
Still don't understand: Is velox necessary to run my plugin through RPC?
I have:
client.php
```php
$rpc = new Goridge\RPC\RPC(
    Goridge\Relay::create('tcp://127.0.0.1:6006')
);
// ...
$router = new \League\Route\Router();
$router->get('/on_purpose_test', static function (ServerRequestInterface $request) use (&$psr7, &$rpc): Response {
    return new Response(200, [], $rpc->call("App.Hi", "eugene"));
});
```
rr (binary)
requests.go
```go
package main

import (
	"fmt"
	"log"
	"net"
	"net/rpc"

	goridgeRpc "github.com/spiral/goridge/v4/pkg/rpc"
)

// App sample
type App struct{}

// Hi returns greeting message.
func (a *App) Hi(name string, r *string) error {
	*r = fmt.Sprintf("Hello, %s!", name)
	return nil
}

func main() {
	ln, err := net.Listen("tcp", ":6006")
	if err != nil {
		panic(err)
	}

	err = rpc.Register(new(App))
	if err != nil {
		panic(err)
	}
	log.Printf("started")

	for {
		conn, err := ln.Accept()
		if err != nil {
			continue
		}

		log.Printf("new connection %+v", conn)
		go rpc.ServeCodec(goridgeRpc.NewCodec(conn))
	}
}
```

Just to check if this would work. Should it work? Or am I missing smthn?

**Comment by rustatian** [comment]
This is a standalone server (which should be started separately), not a plugin.

> Just to check if this would work.

- `go run .`
- `php client.php`

**Reply by PakToPack** [reply]
Idk what's wrong, it won't start. vendor folder is for composer files.

Also I have a question:
**Would this script work properly and fast, when I would have like 2-3k rps?** Like 800 workers continuously making rpc calls to it. Would it handle that with `go run requests.go`?
```bash
[root@eu roadrunner]# go run requests.go
go: finding module for package github.com/spiral/goridge/v4/pkg/rpc
requests.go:9:2: cannot query module due to -mod=vendor
        (Go version in go.mod is at least 1.14 and vendor directory exists.)
```

```bash
drwxr-xr-x  5 root root     4096 Jun 20 18:24 .
drwxr-xr-x  6 root root       70 Jun 14 10:58 ..
-rwxr-xr-x  1 root root       68 Jun 14 12:09 .gitignore
-rw-r--r--  1 root root     1213 Jun 20 17:51 .rr.yaml
-rw-r--r--  1 root root     1069 Jun 20 14:31 LICENSE
-rw-r--r--  1 root root     1652 Jun 20 14:31 README.md
-rw-r--r--  1 root root     2148 Jun 14 19:32 RemoteDB.php
-rw-r--r--  1 root root     9807 Jun 14 12:09 app.php
-rw-r--r--  1 root root     6996 Jun 20 17:51 clean.php
-rw-r--r--  1 root root      439 Jun 14 19:32 composer.json
-rwxr-xr-x  1 root root    99884 Jun 14 11:05 composer.lock
-rw-r--r--  1 root root      147 Jun 20 17:41 config.go
-rw-r--r--  1 root root       23 Jun 20 18:24 go.mod
drwxr-xr-x  2 root root      150 Jun 14 17:27 logs
-rw-r--r--  1 root root     1786 Jun 20 17:41 makeRequest.go
-rw-r--r--  1 root root      613 Jun 20 18:24 requests.go
-rwxr-xr-x  1 root root 48488448 Jun 20 18:14 rr
drwxr-xr-x  4 root root       31 Jun 14 11:18 rr_cache_dir
-rw-r--r--  1 root root     9872 Jun 15 15:26 test.php
-rw-r--r--  1 root root     3109 Jun 20 18:15 velox.toml
drwxr-xr-x 13 root root      181 Jun 14 11:05 vendor
-rwxr-xr-x  1 root root  9404416 Jun 20 14:31 vx
```

**Reply by PakToPack** [reply]
> Idk what's wrong, it won't start. vendor folder is for composer files.
> 
> 
> 
> Also I have a question:
> 
> **Would this script work properly and fast, when I would have like 2-3k rps?** Like 800 workers continuously making rpc calls to it. Would it handle that with `go run requests.go`?
> 
> ```bash
> 
> [root@eu roadrunner]# go run requests.go
> 
> go: finding module for package github.com/spiral/goridge/v4/pkg/rpc
> 
> requests.go:9:2: cannot query module due to -mod=vendor
> 
>         (Go version in go.mod is at least 1.14 and vendor directory exists.)
> 
> ```
> 
> 
> 
> ```bash
> 
> drwxr-xr-x  5 root root     4096 Jun 20 18:24 .
> 
> drwxr-xr-x  6 root root       70 Jun 14 10:58 ..
> 
> -rwxr-xr-x  1 root root       68 Jun 14 12:09 .gitignore
> 
> -rw-r--r--  1 root root     1213 Jun 20 17:51 .rr.yaml
> 
> -rw-r--r--  1 root root     1069 Jun 20 14:31 LICENSE
> 
> -rw-r--r--  1 root root     1652 Jun 20 14:31 README.md
> 
> -rw-r--r--  1 root root     2148 Jun 14 19:32 RemoteDB.php
> 
> -rw-r--r--  1 root root     9807 Jun 14 12:09 app.php
> 
> -rw-r--r--  1 root root     6996 Jun 20 17:51 clean.php
> 
> -rw-r--r--  1 root root      439 Jun 14 19:32 composer.json
> 
> -rwxr-xr-x  1 root root    99884 Jun 14 11:05 composer.lock
> 
> -rw-r--r--  1 root root      147 Jun 20 17:41 config.go
> 
> -rw-r--r--  1 root root       23 Jun 20 18:24 go.mod
> 
> drwxr-xr-x  2 root root      150 Jun 14 17:27 logs
> 
> -rw-r--r--  1 root root     1786 Jun 20 17:41 makeRequest.go
> 
> -rw-r--r--  1 root root      613 Jun 20 18:24 requests.go
> 
> -rwxr-xr-x  1 root root 48488448 Jun 20 18:14 rr
> 
> drwxr-xr-x  4 root root       31 Jun 14 11:18 rr_cache_dir
> 
> -rw-r--r--  1 root root     9872 Jun 15 15:26 test.php
> 
> -rw-r--r--  1 root root     3109 Jun 20 18:15 velox.toml
> 
> drwxr-xr-x 13 root root      181 Jun 14 11:05 vendor
> 
> -rwxr-xr-x  1 root root  9404416 Jun 20 14:31 vx
> 
> ```

Also I don't really know how to properly implement this.

The idea was:
1. Incoming get request with data
2. Checking data using PHP
3. Sending a GET requests to partners sites through Go script(or plugin)
4. Response from Go to PHP with outgoing GET Request result

And this should hold 2-3k rpm of load.

**Reply by rustatian** [reply]
Your PHP `vendor` folder conflicts with Go's. Go and PHP should remain in separate folders to avoid conflicts OR you can rename the PHP `vendor` folder in `composer.json`.

**Reply by PakToPack** [reply]
Hello!
I get this error:
```
Error: Error 'illegal base64 data at input byte 0'
```
The code is:
```php
$router->get('/rpc', static function (ServerRequestInterface $request) use (&$psr7, &$rpc): Response {
    $arr = [
        "https://google.com/",
        "https://1.1.1.1",
        "http://eu.adsy.ink"
    ];
    return new Response(200, [], $rpc->call("App.MakeRequest", json_encode($arr)));
});
```

Go code:
```go
func (a *App) MakeRequest(jsonData string, r *string) error {
	var jsonUrls []string

	jsonErr := json.Unmarshal([]byte(jsonData), &jsonUrls)
	if jsonErr != nil {
		log.Fatal(jsonErr)
	}

	var testDecode string

	for _, u := range jsonUrls {
		u := u
		testDecode += "\n" + u
	}
	*r = fmt.Sprintf(testDecode)
	return nil
}
```

**Reply by rustatian** [reply]
Could you please show the full stack trace?

**Reply by PakToPack** [reply]
I did smthn wrong in method implementation. After copying Hi method and renaming it - all works fine.

**Reply by PakToPack** [reply]
Rewrote my GoLang code to this(still doesn't work):
```go
package main

import (
	"encoding/json"
	"fmt"
	_ "github.com/roadrunner-server/api/v2/plugins/config"
	_ "github.com/roadrunner-server/errors"
	goridgeRpc "github.com/roadrunner-server/goridge/v3/pkg/rpc"
	_ "go.uber.org/zap"
	"io"
	"log"
	"net"
	"net/http"
	"net/rpc"
	"sync"
	"time"
)

// App sample
type App struct{}

var netClient = &http.Client{
	Timeout: time.Millisecond * 120,
}

func main() {
	ln, err := net.Listen("tcp", ":6969")
	if err != nil {
		panic(err)
	}

	err = rpc.Register(new(App))
	if err != nil {
		panic(err)
	}
	log.Printf("started")

	for {
		conn, err := ln.Accept()
		if err != nil {
			continue
		}

		log.Printf("new connection %+v", conn)
		go rpc.ServeCodec(goridgeRpc.NewCodec(conn))
	}
}

// Hi returns greeting message.
func (a *App) Hi(name string, r *string) error {
	*r = fmt.Sprintf("Hello, %s!", name)
	return nil
}

func (a *App) MakeRequest(data string, r *string) error {
	start := time.Now()
	var jsonUrls []string

	jsonErr := json.Unmarshal([]byte(data), &jsonUrls)
	if jsonErr != nil {
		log.Fatal(jsonErr)
	}
	var testDecode string
	for _, u := range jsonUrls {
		u := u
		testDecode += "\n" + u
	}

	var wg sync.WaitGroup
	ch := make(chan string)
	for _, u := range jsonUrls {
		wg.Add(1)
		go doReq(u, ch, &wg)
	}
	close(ch)
	wg.Wait()

	for range jsonUrls {
		testDecode += <-ch + "\n"
	}
	*r = fmt.Sprintf("%.2fs elapsed, count of links: %d\n\nReceived data:\n%s", time.Since(start).Seconds(), len(jsonUrls), testDecode)
	return nil
}

func doReq(url string, ch chan<- string, wg *sync.WaitGroup) {
	defer wg.Done()
	resp, err := netClient.Get(url)
	if err != nil {
		return
	}
	body, err := io.ReadAll(resp.Body)
	ch <- fmt.Sprintf("%s \n\n", string(body))
}
```
But now I'm getting(Error, but at least not overtime):
```
Line: 149
Error: Unable to read frame header: Success
```
What's interesting: I don't have a line 149 in php code, that handles rpc testing route. Also I don't have 149 line in GoLang script

![image](https://github.com/roadrunner-server/roadrunner/assets/37382014/a6efcb21-69fb-43d6-ab42-46a660baf2db)

![image](https://github.com/roadrunner-server/roadrunner/assets/37382014/e6116f1c-a32e-459d-84cc-783b3bd3c8db)

**Reply by PakToPack** [reply]
> Could you please show the full stack trace?

Hope you'd help, cuz I'm not a Go programmer 😅

**Reply by rustatian** [reply]
Have you read the tutorials I sent you?

**Reply by PakToPack** [reply]
> Have you read the tutorials I sent you?

Yep, kinda. I read those that mention http and goroutines. But idk 'bout this errors

**Reply by PakToPack** [reply]
> Have you read the tutorials I sent you?

Hello. I double-checked the tutorials you sent me and I didn't see an issue. But there is an issue obviously. Can you please help? (Even just where to look at, cuz idk)

**Reply by rustatian** [reply]
Send me a full stack trace of the panic/error.

**Reply by PakToPack** [reply]
Ok, I get this
```bash
Spiral\Goridge\Exception\HeaderException: Unable to read frame header: Success in /var/www/roadrunner/vendor/spiral/goridge/src/SocketRelay.php:149
Stack trace:
#0 /var/www/roadrunner/vendor/spiral/goridge/src/RPC/RPC.php(63): Spiral\Goridge\SocketRelay->waitFrame()
#1 /var/www/roadrunner/clean.php(41): Spiral\Goridge\RPC\RPC->call()
#2 /var/www/roadrunner/vendor/league/route/src/Strategy/ApplicationStrategy.php(48): {closure}()
#3 /var/www/roadrunner/vendor/league/route/src/Route.php(167): League\Route\Strategy\ApplicationStrategy->invokeRouteCallable()
#4 /var/www/roadrunner/vendor/league/route/src/Dispatcher.php(59): League\Route\Route->process()
#5 /var/www/roadrunner/vendor/league/route/src/Strategy/ApplicationStrategy.php(37): League\Route\Dispatcher->handle()
#6 /var/www/roadrunner/vendor/league/route/src/Dispatcher.php(59): Psr\Http\Server\MiddlewareInterface@anonymous->process()
#7 /var/www/roadrunner/vendor/league/route/src/Dispatcher.php(53): League\Route\Dispatcher->handle()
#8 /var/www/roadrunner/vendor/league/route/src/Router.php(111): League\Route\Dispatcher->dispatchRequest()
#9 /var/www/roadrunner/clean.php(236): League\Route\Router->dispatch()
#10 {main}
```

**Reply by rustatian** [reply]
The error comes from the PHP, not from Go. I guess that you sent a improperly formatted message. I previously sent you repo where you can have a look for the examples: `roadrunner-php/jobs`.

**Reply by PakToPack** [reply]
It comes from go. This is my supervisor log for GoLang script startup:
```bash
2023-06-23 13:02:23,403 INFO exited: golang_server (exit status 2; expected)
2023-06-23 13:02:24,419 INFO spawned: 'golang_server' with pid 20441
```
The message itself can be received. If I'd remove this code:
```go
var wg sync.WaitGroup
	ch := make(chan string)
	for _, u := range jsonUrls {
		wg.Add(1)
		go doReq(u, ch, &wg)
	}
	close(ch)
	wg.Wait()

	for range jsonUrls {
		testDecode += <-ch + "\n"
	}
```
I'll get this:
![image](https://github.com/roadrunner-server/roadrunner/assets/37382014/94c13cbb-ec3a-4fac-a10a-1950e5c14cd0)

The code itself would look like this(and it works perfectly fine):
```go
func (a *App) MakeRequest(data string, r *string) error {
	start := time.Now()
	var jsonUrls []string

	jsonErr := json.Unmarshal([]byte(data), &jsonUrls)
	if jsonErr != nil {
		log.Fatal(jsonErr)
	}
	var testDecode string
	for _, u := range jsonUrls {
		u := u
		testDecode += "\n" + u
	}

	*r = fmt.Sprintf("%.2fs elapsed, count of links: %d\n\nReceived data:\n%s", time.Since(start).Seconds(), len(jsonUrls), testDecode)
	return nil
}
```

So I guess the problem is in that waitgroup code or this one:
```go
func doReq(url string, ch chan<- string, wg *sync.WaitGroup) {
	defer wg.Done()
	resp, err := netClient.Get(url)
	if err != nil {
		return
	}
	defer func(Body io.ReadCloser) {
		err := Body.Close()
		if err != nil {
			return
		}
	}(resp.Body)
	body, err := io.ReadAll(resp.Body)
	ch <- fmt.Sprintf("%s \n\n", string(body))
}
```

**Reply by PakToPack** [reply]
> The error comes from the PHP, not from Go. I guess that you sent a improperly formatted message. I previously sent you repo where you can have a look for the examples: `roadrunner-php/jobs`.

Maybe you can trace, whats wrong? I can't even with tutorials

**Reply by rustatian** [reply]
I sent you an example on how our libraries work. How to perform an `rpc` call from PHP. Look at the `roadrunner-php/jobs` package.

**Reply by PakToPack** [reply]
> I sent you an example on how our libraries work. How to perform an `rpc` call from PHP. Look at the `roadrunner-php/jobs` package.

I did read it now. But I don't need a job to call the site. I need to provide a client, that sent a request to us with a special responses from different web sites.

How can I do that with a Job? If it is actually used to queue the task, not to make it right away?

**Reply by rustatian** [reply]
Since you have an issue with your Go code, not with the RPC calls (or RR), please wait, until I would have a time to help you. Thanks 👍🏻

**Reply by PakToPack** [reply]
> Since you have an issue with your Go code, not with the RPC calls (or RR), please wait, until I would have a time to help you. Thanks 👍🏻 

No problem. I would wait, Appreciate your help 🥲

**Reply by PakToPack** [reply]
> Since you have an issue with your Go code, not with the RPC calls (or RR), please wait, until I would have a time to help you. Thanks 👍🏻 

Hello, did you check that? I'm tryin' my best, but changed the solution for a short-term now

**Reply by rustatian** [reply]
Does the single thread version work?

**Reply by PakToPack** [reply]
> Does the single thread version work?

Yes. But I need concurrency. If I make single HTTP request - all good.
When I make 400 rps - crash.

Now the code above crushing even without the load, so I guess I made a mistake in concurrency.

**Reply by rustatian** [reply]
So, first of all, run your code with `-race` flag. I suspect, that `netClient` can't be global or should be protected by mutex.
Second - always, when you asking something or getting an error - **show the full stack trace** (if applicable).
